### ISSUE TITLE
Compatibility with Nextcloud 14 [$190]

### ISSUE THREAD

**Issue by juliushaertl** [nextcloud/passman#491]
Hi @brantje,

as you may have already seen, we have entered the beta phase for the release of Nextcloud 14. We try to keep up compatibility with older apps, but in some cases apps need some adjustments to properly work on new Nextcloud releases. With Nextcloud 14 there have been quite some changes in that regard:

You can find a list of all critical changes for app developers and admins at https://github.com/nextcloud/server/issues/7827 You can run `occ app:check-code` to validate that there are no issues with deprecated or private message calls. 

One critical change is the overall restructuring of the main template (https://github.com/nextcloud/server/pull/9982) that might require some adjustments in your app.

We would be very pleased if you can make your app compatible with Nextcloud 14 and release a new version to the app store, so that users can properly test it and the upgrade experience to Nextcloud 14 will be even better than before.

Feel free to ask, if you have any questions regarding making your app compatible with Nextcloud 14. 

Thanks for being part of the Nextcloud community and for all the effort you put into providing this app :rocket:

Btw. you might have seen that starting August 25, the Nextcloud Conference takes place at the familiar TU Berlin. (https://nextcloud.com/conf/)! We'd like all contributors to be there and that most certainly includes you. You might already have registered but if you haven't - please consider joining us!

**Comment by vwbusguy** [comment]
Nextcloud 14 was announced today and released RC, but Passman gets disabled on upgrade as of now.

**Comment by Kourai** [comment]
Any news on compatibility with Nextcloud 14 ?

**Comment by vwbusguy** [comment]
Nextcloud 14 was tagged for GA release today but sadly no compatibility update to passman yet.

**Comment by bstin** [comment]
Yes, to flesh out what @Heidistein said, you can force to work by modifying the file: 
`/nextcloud/apps/passman/appinfo/info.xml`
and change the line 
`<nextcloud min-version="9" max-version="13" />`
to 
`<nextcloud min-version="9" max-version="14" />`
It will work, although the rendering is little off. But you can at least access passwords this way until a final version is released.

**Comment by Putdeksel** [comment]
Not sure if this should be here but @bstin thank you, I was about to downgrade. I do have trouble finding that file though, it seems there is no passman folder in my /nextcloud/apps. Do you have any suggestions on where to look?

**Comment by tcassaert** [comment]
@Putdeksel You can find that file in your webroot, which is probably `/var/www/nextcloud`. From  there you can navigate to `apps/passman/appinfo/info.xml`.

**Comment by Putdeksel** [comment]
Thank you for your reply @tcassaert. It seems that because I use the Techandme.se VM and their script to update instead of the traditional method in Nextcloud, unsupported apps are removed during the update. I've tried to copy the apps/passman/ folder that I found in the backup directory, into the var/www//nextcloud/apps directory and editing the info.xml file as suggested by @bstin but without any luck. Guess I'll have to restore my backup :( Thanks anyway!

**Comment by bstin** [comment]
@Putdeksel 

Just to add, after I edited info.xml I had to manually enable pasmman it in the list of apps. After you restored apps/passman directory back into the correct location, do you see it in list of apps?

**Comment by himpierre** [comment]
Is there an official word from the maintainer? Will there be a nextcloud 14 version?

**Comment by tcassaert** [comment]
From how it looks now, I don't think that will happen very soon. The latest "real" commit, apart from translation commits, already goes back to the first of April 2018.

**Comment by himpierre** [comment]
Oh dear. I guess i have to switch to another password manager.

**Comment by vwbusguy** [comment]
Well, we could always fork it.  I'd love to see this continue to be maintained.  I've enjoyed the user experience so far.  I'm willing to do some coding work but the maintainer seems MIA.  I'm not sure it's worth putting in a PR if the maintainer won't even acknowledge stuff like this.

**Comment by vwbusguy** [comment]
It looks like @brantje is active on GitHub - even today - just not here..

**Comment by vwbusguy** [comment]
Looks like there was an update the the web extension today, so maybe there's hope? https://github.com/nextcloud/passman-webextension

**Comment by synthead** [comment]
If you need to get by, Passman still kinda works with a broken layout as-is on Nextcloud 14.  Find where Passman is installed, then edit `appinfo/info.xml`.  In there, you'll find this line:

```xml
        <nextcloud min-version="9" max-version="13" />
```

Change this to:

```xml
        <nextcloud min-version="9" max-version="14" />
```

This also seems to make the Passman API work, so the Android Passman app works just fine like this.

**Comment by Bubu** [comment]
Please do open a merge request! People can still apply it manually on their instances.

**Comment by vwbusguy** [comment]
Actually, I started doing the PR to find that the version has been bumped in this repo 6 months ago - it was just never pushed as a release to NextCloud:
https://github.com/nextcloud/passman/blob/master/appinfo/info.xml#L46

**Comment by vwbusguy** [comment]
#497 is ultimately a blocker for this, since none of the PRs are passing tests since it's trying to test NC14 against PHP 5.x.

**Comment by vwbusguy** [comment]
It looks like all @brantje would need to do is tag a release off of master and NC14 would work again without the hack.

**Comment by glukose** [comment]
Hello,

I've also changed my version number in the file and was able to enable back the app in the admin panel.
However I can't access to my vaults through the web interface (password incorrect). The Firefox extension does open my vault but all the fields are set to 'false'. Labels of the credentials are not changed.
I successfully registered a new password through the Firefox extension though. 

Anyone got the same problem ? 
Is it possible I have deleted a file in my data/ directory of importance ?  (I cleaned up files_thrashbin before the migration)

Cheers

**Comment by Oliv4945** [comment]
@glukose I used [install from Git](https://github.com/nextcloud/passman/wiki/Installation) instructions, works well without editing anything.

Nevertheless I put some money in the bountysource, I hope it will motivate developers (@brantje  ?) to continue & debug this useful app :)

**Comment by bstin** [comment]
Hi @glukose 

I have not had that problem. My password works, fields are fine, and Firefox extension works. For reference here is a listing of all files under my `/var/www/nextcloud/apps/passman` directory so you can compare against yours. These include the md5, so copy to your server where you should be able to theoretically test by running 
`cd /var/www/nextcloud/apps/passman` _(or wherever your apps live)_

`md5sum -c filename_md5.txt`

Regards

Files:
[passman_files_md5.txt](https://github.com/nextcloud/passman/files/2387082/passman_files_md5.txt)

**Comment by glukose** [comment]
Thanks @Oliv4945 & @bstin 
The md5 was correct. I also tried the "git install" to get 2.1.5. None of these worked, same problem.

Do you know were my vaults keys are store ? In the database table "oc_passman_vaults" and field "private_sharing_key" ?

I'll try to restore to an older nextcloud version, but I don't think my chances to success are very high.

**Comment by happyreacer** [comment]
Nothing reaction from the developer... :-(
It is so bad for all user

**Comment by seraphyn** [comment]
Vacancy ?
Hope so

**Comment by twixwix** [comment]
Is there any movement with this update? We need a new commit @brantje

**Comment by seraphyn** [comment]
We have now August to October @brantje .
Did you left the project?
Will the behavior the same, when we have a security breach?

Is something willing here to fork and go on with the project?

**Comment by florealcab** [comment]
I am afraid this plugins seems to not be maintained anymore... I hope a new maintainer will take the relay, or the actual maintainer will release a new version to fix it soon...

**Comment by newhinton** [comment]
@jospoortvliet @juliushaertl @vwbusguy @brantje @animalillo 

I am currently in the process of reworking the UI for NC14.
See [Branch](https://github.com/nextcloud/passman/tree/fix_layout_for_14)

It will need some more time, but i will fix it sometime around next week. I will also take a peak at the most pressing bugs which have been reported, but this will take additional time.

And now the big BUT: what if @brantje wont show up? even if i just merge my PR and ignore good manners, the update still needs to get uploaded to nc, but this wont be possible without the original maintainers. 
What do you suggest in this case?

**Comment by Heidistein** [comment]
Doing great stuff I see.
What if @brantje won't show up. Well, you already started a fork. Are you willing to become a new maintainer?

I think a way could be found to merge your fork without the original creator into nextcloud. I really do insist that the credits as who created stuff are keps sane and correct.

If you will not want to be a maintainer, well... I'll step up. It wont be ideal, far from. But I do like this plugin more than that.

@brantje :
Hey man! Echt, je hebt iets gemaakt dat vet veel mensen tof vinden. we willen je code blijven gebruiken.
Er staat ~130 euro voor je klaar. Mensen geven vrijwillig geld voor je werk, en terecht vind ik. Zou je wakker willen worden er er iets van vinden?

**Comment by newhinton** [comment]
I have fixed most of the bugs, even CI is running fine now.
(needed to disable two of the tests, but i suspect that this was an error with the used owncloud/nextcloud instance, not passman. should be fine imo)

**Comment by cpranzl** [comment]
Just sent you $5 via PayPal to motivate you.

**Comment by e-alfred** [comment]
@newhinton Maybe you want to join the Telegram channel to get into contact faster? https://t.me/passman_general

**Comment by MorrisJobke** [comment]
@newhinton just talked with me and I want to give an update here in the public as well. As it seems that the current maintainer is not active anymore I proposed to revoke the existing app certificate and generate a new one (I would wait ~2 weeks for @brantje to put a veto to this revocation).

Then @newhinton has a valid key and I can add him as maintainer to apps.nextcloud.com so that he can publish updates. Until then you all could work here to finish the NC14 compatibility. As there was no (or at least very little) reaction by the current maintainer I would give @newhinton for now write permissions to this repo. In case of problems we could revert to the current master version anyways (yay to version control systems 😉).

Does that sound like a plan?

**Comment by MorrisJobke** [comment]
Just to make one thing clear: @brantje, I don't want to overrule you - I just want to give the community the opportunity to continue your work in a good fashion. It's fine if you don't have any time anymore. Thanks for your contributions anyways. If you are not fine with this we could do it by a real fork as well, but from a users perspective it would be nice to just continue here and roll out updates.

**Comment by joergmschulz** [comment]
This is great!

Jörg M. Schulz

**Comment by AndyScherzinger** [comment]
I posted this issue on the passman Telegram group and pinged @brantje explicitly. So let's see.

**Comment by brantje** [comment]
Hi,
Sorry for long time away, real life is really busy.
I will try to find some time upcoming weeks for passman to update it.
I will not however give permission to revoke my app signing key's.
Passman is kinda my baby, hope you understand it.

**Comment by AndyScherzinger** [comment]
@brantje how about a collaboration for Nc14 since @newhinton seems to have done most of the necessary work and it seems to be a shame to do this work twice especially when the day life keeps you busy like it also keeps me busy these days.

**Comment by brantje** [comment]
Ye saw the prs, will review them asap =)

**Comment by animalillo** [comment]
I don't think it's fair to give ultimatums on free opensource projects.
Nobody is paying us to maintain passman, so sorry if we decide to work and earn money to be able to pay for a house.

If nextcloud is willing to pay me for a full time job maintaining passman I will gladly consider this, I love this project and I feel insulted by this kind of behavior. Very unprofessional from your side, in my opinion at least.

I don't think that neither me nor @brantje are willing to leave this project this way.

**Comment by seraphyn** [comment]
@animalillo Sorry for getting things wrong, but since August some peoples looking for an answere.
I waited like any other just for this answere for more than 8 weeks. Patches in the line, bounty etc
Now I just get an very pissed answere how unfair we are, that we talked about forking the project. We are user of this project and we're depend on it.

As I asked, see  [above](https://github.com/nextcloud/passman/issues/491#issuecomment-424039639), will this be the same timeline when a security breach is found?
From this time on, it tooks more than two weeks to get an answere.

Wouldn't it be great to say, sure of course, we took another one in the boat to help us reviewing patches and in some cases upload them?

This will need to save time and trouble on bouth sides.
You for your project, and of course the users, the ones who trusted you for securing their data.
Noone will kick an busy developer out of his project. But before it dies and no answere is given, a fork will be the normal way. Isn't it like that?

**Comment by himpierre** [comment]
> Isn't it like that?

Yes it is. And considering the answer from @animalillo it might be a good idea to fork. Better sooner than later. I myself do not feel i can rely on that software right now.

**Comment by newhinton** [comment]
@seraphyn @animalillo Hey everybody, calm down :D we're all friends here ;)

@seraphyn i understand that you are upset, but now that they are back, im sure we will see progress on this project

but @animalillo you have to agree that a project which hasn't recieved any commits for nearly 6 month's and no contact to the main devs at all is suspiciously silent, and since it broke completely for nc14 with no reaction from you guys, it was just the sensible thing to assume that you have abandoned the project. It's great news that you didn't and i would be really glad to help you out, but if that wouldn't be the case, somebody could take over the project and keep it alive, and no harm would be done. 


Also nobody was going to "throw you out" or "make you leave" the project, it just _seemed_ that you didn't want to continue work, which is fine by any means.

But people want to improve stuff they may use daily, and that is fine by any means too. 

Also there was no 'Ultimatum' in a sense that you are out then. It was just a timeframe for checking with you guys to find out if you are still there and to prevent any throwing out or upsetting you

tl;dr
Pauses are fine, no harm done, but lets not blame each other and improve passman from now on.

**Comment by MorrisJobke** [comment]
> Also nobody was going to "throw you out" or "make you leave" the project, it just _seemed_ that you didn't want to continue work, which is fine by any means.

@brantje @animalillo I also didn't want to kick you out. That's the reason why I asked for your part of the story. And good to hear from you. 👍 @newhinton just want to help keep this project up and running. And having helping hands should be the way to go. We are a community and should work together towards a better ecosystem.

Sorry if you understood me wrong.

**Comment by brantje** [comment]
I'm getting pissed, people demanding things for stuff we do in our free time!
This is not how opensource is meant to be. 
I understand that people depend on the app, as do i, using it daily ;).

`As I asked, see above, will this be the same timeline when a security breach is found?`
Tell me a breach in passman =)

`What's your timeline in security related software?`
Wrong question, you seem to forget the project is ran by volunteers who give away their time for free!
As said before, if people want quick updates + patches, nextcloud should hire @animalillo and me so we can work full time on the app. If they don't want to, then well, things have to wait till we have more time.

```Wouldn't it be great to say, sure of course, we took another one in the boat to help us reviewing patches and in some cases upload them```
Maybe, that's up to @animalillo and me to decide, not Nextcloud.

I really don't appreciate the way how things went. Nextcloud has my personal contact details, telegram group is open for everyone, yet nobody joined / e-mailed me.

```Also nobody was going to "throw you out" or "make you leave" the project, it just seemed that you didn't want to continue work, which is fine by any means.```
Well, assumptions are the mother of all f* ups. 
Feel free to send me a telegram message next time.

**Comment by newhinton** [comment]
@brantje Actually, i wrote an email, a week ago. And to be honest, you did not answer to any of the issues and mentions which have been popping up over the last few month's,so the only thing we could do was assuming stuff.

**Comment by Oliv4945** [comment]
@brantje @animalillo: Please avoid assuming what everyone thinks because some people expressed their frustrations in a way other might also disapprove.

On my side I am a huge fan of your work on Passman, thank you for that ! But given your activity on this git repo I was also assuming that you might not want to continue it anymore, and started to look for maintained alternatives.
If I knew that your preferred way of communication is Telegram, I would have give a try some times ago, and it might have avoid all this misunderstanding: maybe you should write it on the main readme :-)

**Comment by brantje** [comment]
@newhinton sorry if i sounded a bit rude that was not my intention. Could you please join: https://t.me/passman_general ?

**Comment by MorrisJobke** [comment]
Just to keep the people over here updated: it seems that some comments were a bit rougher than expected and we realized that we have a common goal. It seems that @newhinton will help @brantje and @animalillo to get Passman compatible with Nextcloud 14 within the near future.

Please also don't judge people too fast on the comments. Sometimes the feelings behind the words in here just get interpret the wrong way. We had a chat in the telegram group to exchange a bit more the feelings of each other. It's just hard to communicate always on point without leaving room for misunderstandings.

Respect each other and be kind to each other. ❤️

**Comment by gt2416** [comment]
Just wanna say thank you for the project @brantje  and @animalillo for passman. 
We want to get it fixed asap cause passman is a CRITICAL piece of software for a user. 
As a community I'm sure we all respect you and love your work, thanks again for your time.

**Comment by Luticus** [comment]
The app isn't controlled by one person, it's controlled by two, now possibly three.  I think we all need to take a step back and realize that pissing off the developers will not get your app updated faster.  I completely understand how it is to develop something or create something and give it away for free only to have people turn around and get all demanding with it.  I too am frustrated that I'm stuck on Nextcloud 13 until this app is updated for 14, but that only goes to show how important this app is.  We need to understand that these guys are not paid developers, they are providing a free service.  In the past the guys here have been really great and even implemented a fantastic feature I requested (vault destruction requests).  While I do wish they would monitor github a bit better (that's where people generally file bug reports) and let us know they are still planning the keep the app alive, I'm certainly not going to demand anything.  If you want the app done faster, increase the bounty, offer a pull request, or offer some other sort of help.  Otherwise, do what I'm doing:  wait patiently.

**Comment by ontherunvaro** [comment]
Question: if I upgraded my Nextcloud version to 14, and passman was disabled, is the stored data now lost?

**Comment by Depado** [comment]
> Question: if I upgraded my Nextcloud version to 14, and passman was disabled, is the stored data now lost?

No. If you updated Nextcloud and want to see your data anyway, you can do the steps that were mentionned earlier.

Modify the file `<path of nextcloud>/apps/passman/appinfo/info.xml` and change the line:
```xml
<nextcloud min-version="9" max-version="13" />
```
To
```xml
<nextcloud min-version="9" max-version="14" />
```

Additionally you may check the permissions (and owner) of the `passman` directory and change it back to what it was before if it changed like this `chown -R www-data:www-data passman`. Check the other app directories if you're not sure about which user it should be owned by. 

The next step is to go to your nextcloud instance, connect, go the your apps and enable Passman again.

**Comment by synthead** [comment]
@ontherunvaro The stored data won't be lost.  I upgraded and had Passman disabled :slightly_smiling_face:

**Comment by himpierre** [comment]
Is there an e.t.a. for the nextcloud 14 ready release?

**Comment by Luticus** [comment]
If you want a good work around, I just upgraded to NC14 because I've been testing the layout fix for 14 on a test server for a while now and I finally deemed it "good enough" to no longer delay my upgrade.  What I did was go to my server, do the nextcloud upgrade to 14.0.3, then once everything was up-to-date, I changed to the apps directory and cloned the repo.  (git clone https://github.com/nextcloud/passman.git)  Then I checked out the fix branch (git checkout fix_layout_for_14).  Finally I chown -R www-data.www-data passman the directory.  The app works decently well, has some of the newer features, and even works with the passman browser plugin and even seems to work with passman alpha on android (which doesn't work that well anyway, alpha and all).  Not as good as the finished deal, but it's a decent stopgap that will ease the pain a bit so they can take their time and get it done well.

**Comment by newhinton** [comment]
Hey guys! We are almost ready to publish passman 2.2.0 for Nextcloud 14 and 15! The Update includes many fixes to bugs reported by you, and i also overhauled the userinterface! Stay tuned for the update in the close future!

**Comment by JB1985** [comment]
Can you please update the AppStore? -> https://apps.nextcloud.com/apps/passman

**Comment by newhinton** [comment]
@everyone

We updated the app, and the github release is here: [Github](https://github.com/nextcloud/passman/releases)

Also the appstore has been updated: [Appstore](https://apps.nextcloud.com/apps/passman)

All your instances should get the update notification in the next time!

**Comment by AndyScherzinger** [comment]
Nice work everybody! 🚀

**Comment by happyreacer** [comment]
thanks a lot for all your work!

**Comment by Jab2870** [comment]
Thanks everyone for your work.

Is there anything special we need to do if we have manually modified the `info.xml` file?

**Comment by newhinton** [comment]
@Jab2870 to my knowledge, you only need to do something if you now have a version equal or higher to 2.2.1.
if you have checked out the fix_nc14 branch, you should be good to go ;)

but it may take some time that your instance registers the update

**Comment by newhinton** [comment]
@Jab2870

I have had problems with the update, i needes to move the git version to a different foldername to allow the proper app to install

**Comment by muckinger** [comment]
@newhinton 
Thx for your work on this. But i have updated now to 2.2.1 via UI and still get the message "Server version 13 or lower is required." We are on Nextcloud 14.0.1. Any clue?

**Comment by muckinger** [comment]
@Bubu Thx, works now!
